### PR TITLE
Adds an initial typescript typing

### PR DESCRIPTION
**Pull request by allen-munsch**
```
Promise<QueryResultsInterface<HighestCumulativeProfitInterface>>
QueryResultsInterface<SomeCustomInterface>
```

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    type GenericObject = {
+        [names: string]: any;
+    };
+
+    interface QueryResultsInterface<T> {
+        Items: T[];
+        DataScannedInMB: number;
+        QueryCostInUSD: number;
+        EngineExecutionTimeInMillis: number;
+        Count: number;
```
Comment (darcoli) [inline index.d.ts]: should add `S3Location` here

**Comment by alfaproject** [comment]
Is there any hope to get this merged? Even AWS is using TypeScript nowadays, so it would be good to have this (:

**Comment by ghdna** [comment]
Yes, do you think you can update this on the latest code. I merged a bunch of PR and can take this next.

File: index.d.ts (hunk 1 above)
Comment (allen-munsch) [reply index.d.ts:19]: @darcoli added

Comment (allen-munsch) [reply index.d.ts:20]: @darcoli also added `QueryExecutionId`, not quite sure if the typing of string is correct though.

**Comment by allen-munsch** [comment]
@alfaproject @ghdna added two other items to the `QueryResultsInterface` and merged the most current master. Just @ me if there is anything else that needs to be changed on it. For some reason didn't see @darcoli  review on it until your comments showed up.

**Comment by alfaproject** [comment]
@ghdna this should be good to go now, PTAL

File: index.d.ts (hunk 1 above)
Comment (darcoli) [reply index.d.ts:20]: Yes typing is correct

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    interface QueryInterface {
+        sql: string;
+        db: string;
+    }
+
+    type QueryResult = QueryResultsInterface<any>;
+    type QueryFunc = (query: QueryInterface) => Promise<QueryResult>;
```
Comment (darcoli) [inline index.d.ts:29]:
```suggestion
    interface QueryObjectInterface {
        sql: string;
        db: string;
    }
    type DirectQueryString = string;
    type QueryExecutionId = string;

    type OptionalQueryResultsInterface<T> = Partial<QueryResultsInterface<T>> & Pick<QueryResultsInterface<T>, 'QueryExecutionId'>;
    type QueryResult<T> = OptionalQueryResultsInterface<T>;
    type QueryFunc<T> = (query: QueryObjectInterface|DirectQueryString|QueryExecutionId) => Promise<QueryResult<T>>;
```

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+declare module 'athena-express' {
+    import * as aws from 'aws-sdk';
+    interface ConnectionConfigInterface {
+        aws: typeof aws;
+        s3: string;
+        getStats: boolean;
+    }
```
Comment (darcoli) [inline index.d.ts:7]:
```suggestion
        db: string,
        workgroup: string,
        formatJson: boolean,
        retry: number,
        ignoreEmpty: boolean,
        encryption: Record<string, string>,
        skipResults: boolean,
        waitForResults: boolean
    }
```

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    type QueryResult = QueryResultsInterface<any>;
+    type QueryFunc = (query: QueryInterface) => Promise<QueryResult>;
+    interface AthenaExpressInterface {
+        new: (config: ConnectionConfigInterface) => any;
+        query: QueryFunc;
+    }
+
+    class AthenaExpress {
+        new: (config: ConnectionConfigInterface) => any;
+        constructor(config: ConnectionConfigInterface);
```
Comment (darcoli) [inline index.d.ts:37]:
```suggestion
        new: (config: Partial<ConnectionConfigInterface>) => any;
        constructor(config: Partial<ConnectionConfigInterface>);
```

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        sql: string;
+        db: string;
+    }
+
+    type QueryResult = QueryResultsInterface<any>;
+    type QueryFunc = (query: QueryInterface) => Promise<QueryResult>;
+    interface AthenaExpressInterface {
+        new: (config: ConnectionConfigInterface) => any;
+        query: QueryFunc;
+    }
```
Comment (darcoli) [inline index.d.ts:33]:
```suggestion
```

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    interface ConnectionConfigInterface {
+        aws: typeof aws;
+        s3: string;
+        getStats: boolean;
+    }
+
+    type GenericObject = {
+        [names: string]: any;
+    };
+
```
Comment (darcoli) [inline index.d.ts:12]:
```suggestion
```

**Comment by allen-munsch** [comment]
@darcoli they all looked legit to me

**Comment by ghdna** [comment]
This is good to go right? I can spend time this weekend

**Comment by dctrotz** [comment]
I also would love to see this one get merged soon. ;-)

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    type OptionalQueryResultsInterface<T> = Partial<QueryResultsInterface<T>> & Pick<QueryResultsInterface<T>, 'QueryExecutionId'>;
+    type QueryResult<T> = OptionalQueryResultsInterface<T>;
+    type QueryFunc<T> = (query: QueryObjectInterface|DirectQueryString|QueryExecutionId) => Promise<QueryResult<T>>;
+
+    class AthenaExpress {
+        public new: (config: Partial<ConnectionConfigInterface>) => any;
+        public query: QueryFunc;
+        constructor(config: Partial<ConnectionConfigInterface>);
+        query: QueryFunc;
```
[the remark this answers is not in the document]
Comment (dctrotz) [reply index.d.ts:47]: I had to make a few changes to get this to build in my project. See: https://github.com/allen-munsch/athena-express/pull/1

**Pull request merged**